gitlabform/processors/abstract_processor.py

```diff
@@ -89,7 +91,15 @@ def _needs_update(
             entity_in_gitlab.keys()
         )
         for key in keys_on_both_sides:
+            # TODO -> The return from Gitlab has:
+            #  - all the keys, even if their values is None, e.g.
+            #  - keys which are not part of the request API
+            # entity_in_gitlab[deploy_access_levels] -> [{'access_level': 40, 'access_level_description': 'Maintainers', 'user_id': None, 'group_id': None, 'group_inheritance_type': 0}]
+            # != entity_in_configuration[deploy_access_levels] -> [ordereddict([('access_level', 40)])]
```

Comment:
I mean smth like: [5254fbb](https://github.com/gitlabform/gitlabform/pull/423/commits/5254fbb4f79e1760090e783c095ffd103c31959f)

The "entity" returned by Gitlab is a list[dict], and these dicts have more data than what's defined in the cfg (yaml file), basically the key intersection that you do already in `_needs_update`.

I'm afraid of doing that recursively (0: checking if the returned entity is a list, 1: if so iterate it, goto: 0) so I added a custom "analyzer" for a case I know already will be tricky

Though now thinking about it, if this approach makes sense I'll move this function to the base class 🤔

Reply:
Thanks, now I get it.

I also only now have read the docs of this API and I realized that apart from the list of dicts in`deploy_access_levels` it also contains `approval_rules` similar to the multiple approval rules that I am only starting to working on in #388.

Please note that for both of these it would be best to make the GitLabForm syntax slightly easier than the raw API. In particular, we should allow passing usernames and group names instead of their ids. Please see how we have dealt with a similar problem for branch protection in GitLab Premium: https://gitlabform.github.io/gitlabform/reference/protected_branches/#premium-only-features